pkg/kubecost/summaryallocation.go

```diff
@@ -0,0 +1,1155 @@
+			idleAlloc.Name = IdleSuffix
+			aggSet.Insert(idleAlloc)
+		}
+	}
+
+	// TODO
+	// (#) Distribute remaining, undistributed idle
+
+	// Replace the existing set's data with the new, aggregated summary data
+	sas.SummaryAllocations = aggSet.SummaryAllocations
+
+	return nil
+}
+
+func (sas *SummaryAllocationSet) InsertIdleSummaryAllocations(rts map[string]*ResourceTotals, prop AssetProperty) error {
+	if sas == nil {
+		return errors.New("cannot compute idle allocation for nil SummaryAllocationSet")
+	}
+
+	if len(rts) == 0 {
+		return nil
+	}
+
+	// TODO argh avoid copy? Not the worst thing at this size... O(clusters) or O(nodes)
+	idleTotals := make(map[string]*ResourceTotals, len(rts))
+	for key, rt := range rts {
+		idleTotals[key] = &ResourceTotals{
+			Start:   rt.Start,
+			End:     rt.End,
+			CPUCost: rt.CPUCost,
+			GPUCost: rt.GPUCost,
+			RAMCost: rt.RAMCost,
+		}
+	}
+
+	// Subtract allocated costs from resource totals, leaving only the remaining
+	// idle totals for each key (cluster or node).
+	sas.Each(func(name string, sa *SummaryAllocation) {
+		key := sa.Properties.Cluster
+		if prop == AssetNodeProp {
+			key = sa.Properties.Node
+		}
+
+		if _, ok := idleTotals[key]; !ok {
+			// Failed to find totals for the allocation's cluster or node.
+			// (Should never happen.)
+			log.Warningf("InsertIdleSummaryAllocations: failed to find %s: %s", prop, key)
+			return
+		}
+
+		idleTotals[key].CPUCost -= sa.CPUCost
+		idleTotals[key].GPUCost -= sa.GPUCost
+		idleTotals[key].RAMCost -= sa.RAMCost
+	})
+
+	// Turn remaining idle totals into idle allocations and insert them.
+	for key, rt := range idleTotals {
+		idleAlloc := &SummaryAllocation{
+			Name: fmt.Sprintf("%s/%s", key, IdleSuffix),
+			Properties: &AllocationProperties{
+				Cluster: rt.Cluster,
+				Node:    rt.Node,
+			},
+			Start:   rt.Start,
+			End:     rt.End,
+			CPUCost: rt.CPUCost,
+			GPUCost: rt.GPUCost,
+			RAMCost: rt.RAMCost,
+		}
+
+		sas.Insert(idleAlloc)
+	}
+
+	return nil
+}
+
+// Delete removes the allocation with the given name from the set
+func (sas *SummaryAllocationSet) Delete(name string) {
+	if sas == nil {
+		return
+	}
+
+	sas.Lock()
+	defer sas.Unlock()
+
+	delete(sas.externalKeys, name)
+	delete(sas.idleKeys, name)
+	delete(sas.SummaryAllocations, name)
+}
+
+// Each invokes the given function for each SummaryAllocation in the set
+func (sas *SummaryAllocationSet) Each(f func(string, *SummaryAllocation)) {
+	if sas == nil {
+		return
+	}
+
+	for k, a := range sas.SummaryAllocations {
+		f(k, a)
+	}
+}
+
+// IdleAllocations returns a map of the idle allocations in the AllocationSet.
+func (sas *SummaryAllocationSet) IdleAllocations() map[string]*SummaryAllocation {
+	idles := map[string]*SummaryAllocation{}
+
+	if sas == nil || len(sas.SummaryAllocations) == 0 {
+		return idles
+	}
+
+	sas.RLock()
+	defer sas.RUnlock()
+
+	for key := range sas.idleKeys {
+		if sa, ok := sas.SummaryAllocations[key]; ok {
+			idles[key] = sa // TODO Clone()?
```

Comment:
:eyes:

Reply:
Just gonna make this function unexported and not worry about cloning. We only use it one place, and it's not necessary to clone there.